File: lab8/coinchange.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # Initialize memoization table
        memo = [-1] * (amount + 1)
        memo[0] = 0  # Base case: 0 coins needed to make amount 0
        
        # Call the helper function to calculate the minimum coins
        return self.coinChangeHelper(coins, amount, memo)

    def coinChangeHelper(self, coins: List[int], amount: int, memo: List[int]) -> int:
        # If the amount is negative, return -1 (invalid case)
        if amount < 0:
            return -1
        
        # If the result is already computed, return it
        if memo[amount] != -1:
            return memo[amount]

        # Initialize the minimum number of coins as infinity
        minCoins = float("inf")
        
        # Try using each coin
        for coin in coins:
            res = self.coinChangeHelper(coins, amount - coin, memo)
            if res >= 0:  # Valid result
                minCoins = min(minCoins, res + 1)
        
        # If no valid coin combinations, memoize as -1 (not possible)
        memo[amount] = -1 if minCoins == float("inf") else minCoins
        return memo[amount]

File: lab8/CoinChange.java
import java.util.Arrays;

public class CoinChange {

    public int coinChange(int[] coins, int amount) {
        int[] memo = new int[amount + 1];
        Arrays.fill(memo, -1);
        memo[0] = 0;

        return coinChangeHelper(coins, amount, memo);
    }
    
    private int coinChangeHelper(int[] coins, int amount, int[] memo) {
        if (amount < 0) {
            return -1;
        }
        if (memo[amount] != -1) {
            return memo[amount];
        }

        int minCoins = Integer.MAX_VALUE;

        for (int coin : coins) {
            int res = coinChangeHelper(coins, amount - coin, memo);
            if (res >= 0 && res < minCoins) {
                minCoins = res + 1;
            }
        }

        memo[amount] = (minCoins == Integer.MAX_VALUE) ? -1 : minCoins;
        return memo[amount];
    }

    public static void main(String[] args) {
        CoinChange cc = new CoinChange();

        int[] coins1 = {1, 2, 5};
        int amount1 = 11;
        int result1 = cc.coinChange(coins1, amount1);
        System.out.println("Minimum coins required: " + result1);

        int[] coins2 = {2};
        int amount2 = 3;
        int result2 = cc.coinChange(coins2, amount2);
        System.out.println("Minimum coins required: " + result2);

    }
}